packages/eslint-plugin/src/rules/require-await.ts

```diff
@@ -75,13 +84,118 @@ export default createRule({
       ) {
+        let fix: ReportFixFunction | undefined;
+
+        // If the function belongs to a method definition or
+        // property, then the function's range may not include the
+        // `async` keyword and we should look at the parent instead.
+        const nodeWithAsyncKeyword =
+          (node.parent.type === AST_NODE_TYPES.MethodDefinition &&
+            node.parent.value === node) ||
+          (node.parent.type === AST_NODE_TYPES.Property &&
+            node.parent.method &&
+            node.parent.value === node)
+            ? node.parent
+            : node;
+
+        const asyncToken = context.sourceCode.getFirstToken(
+          nodeWithAsyncKeyword,
+          token => token.value === 'async',
+        );
+
+        if (asyncToken) {
+          const nextTokenWithComments = context.sourceCode.getTokenAfter(
+            asyncToken,
+            { includeComments: true },
+          );
+
+          if (nextTokenWithComments) {
+            const asyncRange: Readonly<AST.Range> = [
+              asyncToken.range[0],
+              nextTokenWithComments.range[0],
+            ] as const;
+
+            // Removing the `async` keyword can cause parsing errors if the
+            // current statement is relying on automatic semicolon insertion.
+            // If ASI is currently being used, then we should replace the
+            // `async` keyword with a semicolon.
+            const nextToken = context.sourceCode.getTokenAfter(asyncToken);
+            const addSemiColon =
+              nextToken?.type === AST_TOKEN_TYPES.Punctuator &&
+              (nextToken.value === '[' || nextToken.value === '(') &&
+              (nodeWithAsyncKeyword.type === AST_NODE_TYPES.MethodDefinition ||
+                isStartOfExpressionStatement(nodeWithAsyncKeyword)) &&
+              needsPrecedingSemicolon(context.sourceCode, nodeWithAsyncKeyword);
+
+            const changes = [
+              { range: asyncRange, replacement: addSemiColon ? ';' : '' },
+            ];
+
+            // If there's a return type annotation and it's a
+            // `Promise<T>`, we can also change the return type
+            // annotation to just `T` as part of the suggestion.
+            // Alternatively, if the function is a generator and
+            // the return type annotation is `AsyncGenerator<T>`,
+            // then we can change it to `Generator<T>`.
+            if (node.returnType) {
+              if (scopeInfo.isGen) {
+                if (
+                  hasTypeName(node.returnType.typeAnnotation, 'AsyncGenerator')
+                ) {
+                  changes.unshift({
+                    range: node.returnType.typeAnnotation.typeName.range,
+                    replacement: 'Generator',
+                  });
+                }
+              } else if (
+                hasTypeName(node.returnType.typeAnnotation, 'Promise') &&
+                !!node.returnType.typeAnnotation.typeArguments
```

Comment:
(nit, optional) my personal preference, feel free to ignore
```suggestion
                node.returnType.typeAnnotation.typeArguments != null
```